cpp/src/parquet/encoding_test.cc

```diff
@@ -862,5 +862,215 @@ TEST_F(DictEncoding, CheckDecodeIndicesNoNulls) {
+  const uint8_t data[] = {0xDEU, 0xC0U, 0x37U, 0x13U, 0x11U, 0x22U, 0x33U, 0x44U,
+                          0xAAU, 0xBBU, 0xCCU, 0xDDU, 0x55U, 0x66U, 0x77U, 0x88U};
+  decoder->SetData(sizeof(data) / sizeof(double), data, sizeof(data));
+
+  double values[2] = {.0, .0};
+  const int num_decoded = decoder->Decode(values, 10000);
+  ASSERT_EQ(2, num_decoded);
+
+  uint64_t value_as_uint = *reinterpret_cast<const uint64_t*>(&values[0]);
+  ASSERT_EQ(static_cast<uint64_t>(0x7755CCAA331137DEULL), value_as_uint);
+
+  value_as_uint = *reinterpret_cast<const uint64_t*>(&values[1]);
+  ASSERT_EQ(static_cast<uint64_t>(0x8866DDBB442213C0ULL), value_as_uint);
+}
+
+// Check that the decoder can decode the input in smaller steps.
+TEST(ByteStreamSplitEncodeDecode, DecodeMultipleTimes) {
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  const int num_values = 100;
+  std::vector<uint8_t> data(num_values * 4);
+  for (size_t i = 0; i < data.size(); ++i) {
+    data[i] = static_cast<uint8_t>(i & 0xFFU);
+  }
+  decoder->SetData(num_values, data.data(), num_values * 4);
+
+  const int step = 25;
+  std::vector<float> decoded_data(step);
+  for (int i = 0; i < num_values; i += step) {
+    int num_decoded = decoder->Decode(decoded_data.data(), step);
+    ASSERT_EQ(step, num_decoded);
+    for (int j = 0; j < step; ++j) {
+      const uint32_t assembled_value =
+          static_cast<uint32_t>(data[i + j]) |
+          (static_cast<uint32_t>(data[(i + j) + num_values]) << 8U) |
+          (static_cast<uint32_t>(data[(i + j) + num_values * 2]) << 16U) |
+          (static_cast<uint32_t>(data[(i + j) + num_values * 3]) << 24U);
+      const float assembled_value_as_float =
+          *reinterpret_cast<const float*>(&assembled_value);
+      ASSERT_EQ(assembled_value_as_float, decoded_data[j]);
+    }
+  }
+}
+
+// Check that an encode-decode pipeline produces the original small input.
+// This small-input test is added to ease debugging in case of changes to
+// the encoder/decoder implementation.
+TEST(ByteStreamSplitEncodeDecode, SmallInput) {
+  std::unique_ptr<TypedEncoder<FloatType>> encoder =
+      MakeTypedEncoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+
+  const float data[] = {-166.166f, -0.2566f, .0f, 322.0f, 178888.189f};
+  const int num_values = sizeof(data) / sizeof(data[0U]);
+  encoder->Put(data, num_values);
+
+  std::shared_ptr<Buffer> buffer = encoder->FlushValues();
+
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  decoder->SetData(num_values, buffer->mutable_data(), static_cast<int>(buffer->size()));
+
+  std::vector<float> decoded_data(num_values);
+  int num_decoded_elements = decoder->Decode(decoded_data.data(), num_values);
+  ASSERT_EQ(num_values, num_decoded_elements);
+
+  for (size_t i = 0U; i < decoded_data.size(); ++i) {
+    ASSERT_EQ(data[i], decoded_data[i]);
+  }
```

Comment:
Helper function for checking round trips

Reply:
I will do this in a separate patch.